### PR TITLE
Issue 15

### PR DESCRIPTION
**Pull request by hentai-chan**
+++ Update to the UI +++

Closes issue #15. Note that there are a few things that should be addressed in the future:
- catch invalid user selection in `load_main_menu`
- ensure that `configure_move_set` generates at least on move of type `MoveType::ATTACK` (else you or your opponent won't be able to inflict damage)
- balance difficulty (is `Difficulty::HARD` hard to beat?)

I intent to address the points above in my next PR.

For the first time ever, something noteworthy is happing in `main.cpp`:
1. `check_manifest` seems to be a solid safeguard against missing assets (so far)
2. there's a splash screen, very similar to my hangman game
3. `load_main_menu` enables user customization before the game starts
4. `print_frame` is called for testing purposes, up until then everything should be in working order

I also added a `easy_install.sh` script for Linux users, but you may need to install `apt-get python3-venv` if you never really have used Python before. Should reduce the number of commands one has to type in before the game starts by a fair share :)

### CODE REVIEW

File: src/anim.cpp [hunk 1]
```diff
@@ -34,7 +39,49 @@ std::vector<std::string> gen_healthbar(Pokemon& pkmn)
 	return {label, progressbar, hitpoints};
 }
 
-void print_frame(Pokemon& pkmn1, Pokemon& pkmn2)
+void print_splash_screen(const std::filesystem::path& assets_dir)
+{
+	auto logo = utils::read_file(assets_dir / std::filesystem::path("splashscreen.txt"));
```
Comment (karnkaul) [inline src/anim.cpp]: Just FYI: `fs::path` can be implicitly constructed from string literals too: just `assets_dir / "splashscreen.txt"` will work!

File: src/anim.cpp [hunk 2]
```diff
@@ -34,7 +39,49 @@ std::vector<std::string> gen_healthbar(Pokemon& pkmn)
...
+	std::cout << utils::style(std::accumulate(logo.begin(), logo.end(), std::string("")), utils::Color::YELLOW) << '\n';
+
+	std::cout << '\n' << std::string(19, ' ');
```
Comment (karnkaul) [inline src/anim.cpp]: Just FYI: `cout << setfill(18) << ' '` would avoid needing to construct a string, but length 19 is small enough to fit in SBOs on all compilers anyway.

File: src/anim.cpp [hunk 3]
```diff
@@ -34,7 +39,49 @@ std::vector<std::string> gen_healthbar(Pokemon& pkmn)
...
+
+	for (const char& c : "copyright (c) 2021 cpp-gamedev")
+	{
+		std::cout << c;
+		utils::sleep(100);
```
Comment (karnkaul) [inline src/anim.cpp:52]: Ah so this is the "splash screen" xD

File: src/anim.cpp [hunk 4]
```diff
@@ -34,7 +39,49 @@ std::vector<std::string> gen_healthbar(Pokemon& pkmn)
...
+	// 3. select pokemon
+	std::vector<std::string> names{};
+	names.reserve(pkmns.size());
+	std::for_each(pkmns.begin(), pkmns.end(), [&names](models::Pokemon& pkmn) { names.push_back(pkmn.name); });
+	utils::print_enum_table(names, "pokemons");
+	selection = utils::get_user_input<int>(">>> ");
+	models::Pokemon player = pkmns[selection - 1];
+
+	// 4. remove selection from pkmns, so that player won't fight against his doppelganger
+	pkmns.erase(std::remove_if(pkmns.begin(), pkmns.end(), [player](models::Pokemon pkmn) { return player.id == pkmn.id; }), pkmns.end());
```
Comment (karnkaul) [inline src/anim.cpp:79]: Erase remove idiom, nice!

File: src/anim.cpp [hunk 5]
```diff
@@ -34,7 +39,49 @@ std::vector<std::string> gen_healthbar(Pokemon& pkmn)
...
+
+	return {player, pkmns.size() > 1 ? utils::random_choice(pkmns) : pkmns[0]};
+}
+
+void print_frame(models::Pokemon& pkmn1, models::Pokemon& pkmn2)
```
Comment (karnkaul) [inline src/anim.cpp]: Shouldn't these parameters be `const`?

File: src/main.cpp [hunk 6]
```diff
@@ -1,12 +1,42 @@
...
+		clear_screen();
+
+		print_frame(pkmns[0], pkmns[1]);
 
-	return 0;
+		return EXIT_SUCCESS;
+	}
+	else
+	{
+		std::cerr << "Error!" << '\n';
```
Comment (karnkaul) [inline src/main.cpp]: Might want to print what the error was, so the player can try to fix it.

File: src/utils.cpp [hunk 7]
```diff
@@ -34,6 +38,18 @@ std::string style(std::string text, Color fore, Color back)
 	return ansi_text.append(kt::format_str("{}\033[0m", text));
 }
 
+std::string upper(std::string& str)
```
Comment (karnkaul) [inline src/utils.cpp]: Both `upper` and `lower` transform the string in place, so they shouldn't return anything.

File: src/utils.hpp [hunk 8]
```diff
@@ -24,16 +29,32 @@ T random_range(T min, T max)
 	}
 }
 
+///
+/// \brief Return k random values in range of [min, max] (with repitition)
+///
+template <typename T>
+std::vector<T> random_ranges(T min, T max, std::size_t k)
+{
+	std::vector<T> vector(k);
```
Comment (karnkaul) [inline src/utils.hpp:38]: General suggestion if and when you don't need to return a vector: if you move `k` to be another template parameter instead, you could use an array (or `std::array`) and avoid vector heap allocation.

**Review by karnkaul — APPROVED** [review]

File: src/anim.cpp (hunk 3 above)
Comment (hentai-chan) [reply src/anim.cpp:52]: shy beginnings xD

File: src/utils.hpp (hunk 8 above)
Comment (hentai-chan) [reply src/utils.hpp:38]: Changed `k` to `typename U`, but I'll keep the return type as a vector for now. Will use an array in the future when I move these functions to their own library :)

**Pull request merged**